### DISCUSSION TITLE
Integrate Tibber dynamic pricing for charging costs with geo-fencing

### DISCUSSION THREAD

**Discussion by Hoffis-Welt** [teslamate-org/teslamate#4382] [Ideas]
I propose a new feature in TeslaMate that integrates Tibber's dynamic pricing for calculating charging costs. This feature would allow users to define geo-fences (e.g., home, work, public charging stations) and calculate estimated charging costs based on the current electricity price and the vehicle's location within these geo-fences.

Benefits:

Cost optimization: Users can reduce their charging costs by scheduling charging during off-peak hours.
Planning: Improved planning of charging sessions, especially for long trips or when charging away from home.
Transparency: Visualized charging costs provide better insights into energy consumption.
Integration into existing ecosystem: Seamlessly integrates into TeslaMate's existing features, enhancing its value.
Technical implementation (outline):

Data retrieval: Regularly fetch real-time electricity prices from the Tibber API.
Geo-fencing: Utilize vehicle GPS data and defined geo-fences to determine the current location.
Calculation: Develop an algorithm to calculate charging costs based on electricity prices, vehicle energy consumption, and charging duration.
Visualization: Integrate calculated costs into TeslaMate's user interface.
Notifications: Implement a notification system (e.g., email, push notification) to alert users when charging costs exceed a certain threshold or when electricity prices are particularly low.
Additional considerations:

Flexibility: Allow users to configure different electricity tariffs and providers.
Accuracy: Ensure accurate calculations for reliable decision-making.
Data privacy: Handle user data in compliance with privacy regulations.
Potential extensions:

Integration with other providers: Expand the feature to support other dynamic pricing providers.
Solar integration: Incorporate data from solar installations to optimize charging behavior.
Machine learning: Utilize machine learning to predict future electricity prices and optimize charging schedules.
By integrating Tibber's dynamic pricing, TeslaMate can become an even more powerful tool for Tesla owners, optimizing not only driving behavior but also energy costs.

Would you like me to elaborate on any specific points or add more details to this issue?

Here are some additional points you might want to consider:

User interface: How do you envision the integration into the TeslaMate UI? A new tab, a widget, or overlay?
Error handling: How should the application handle situations where the Tibber API is unavailable or data is corrupted?
User settings: What kind of customization options should be available to users (e.g., preferred currency, notification preferences)?
By providing more specific details, you can help ensure that the developers of TeslaMate have a clear understanding of your proposal and can prioritize it accordingly.

**Comment by brianmay** [comment]
Tibber sounds similar to Amber here in Australia.

Problem is, I image the Tibber API is probably going to be different from the Amber API, which is probably different from other providers in other countries use.

Which isn't to say it can't be done, but ideally needs to be done in a modular way and not tied directly to any one provider.

I would link to the Amber API docs, but looks like you need an Amber account to access them.

**Reply by KrzysztofHajdamowicz** [reply]
How about integration over MQTT or Home Assistant? I have another utility company in Poland that have dynamic tariffs (and there is 6 of them right now), so creating modules inside Teslamate codebase sounds weird for me.
How about publishing current price via MQTT so Teslamate can perform integral operation consumption*price*time?

**Reply by brianmay** [reply]
Interesting idea.

But: with Amber (not sure if Tibber is similar) prices can change at anytime, even after the interval has passed(*). The fact we seem stuck mid way in the transition from 30 minute intervals to 5 minute intervals doesn't help. As a result just publishing current price may not be sufficient.

In my code I currently publish (internally not using mqtt) a list of every known and predicated price for yesterday, today, and tomorrow.

(*) drives me crazy trying to write an algorithm that automatically charges when cheap. The interval often can start of being cheap and end up being expensive.

**Comment by marval** [comment]
I am assuming you have googled this, but you probably missed TeslaMate-Agile - https://github.com/MattJeanes/TeslaMateAgile. Doesn't it fit your needs? You obviously have to spin another docker container, but I imagine this is not an issue.
